### ISSUE TITLE
Access client's subscribed channels in DisconnectEvent

### ISSUE THREAD

**Issue by EmandM** [centrifugal/centrifuge#211]
We are looking to implement behaviour on client Disconnect based on the channels that the client was subscribed to. However, the [`client.close()`](https://github.com/centrifugal/centrifuge/blob/d57e7e003accb56c608c8088af8a5df0fba14b06/client.go#L770) function unsubscribes the client from all channels before calling the disconnect handler.

Is there any way the client's channels can be passed into the `DisconnectEvent`?

**Comment by EmandM** [comment]
We are using centrifuge to pass messages from services through channels to users.

When a client disconnects from centrifuge, we wish to send a notification to the services that the client lost their connection to centrifuge. To do this, we need to have a copy of the channels the client was connected to at the point of the client-side disconnect.

If the client initiates an unsubscribe, that action is considered graceful and no disconnect notification needs to be sent. As the client is maintaining a connection to centrifuge, we do not consider that a disconnect and the service does not need to be notified that the client is gone.

The functionality we are looking for is somewhat like a join/leave notification for the whole system, but we only want to send this notification to the specific channels the client was connected to when they left the system

**Comment by patrick-janeiro** [comment]
Hi @FZambia, 

just to bring a bit of clarity, we are using centrifuge internally somewhat similarly as "pusher". The idea is many of our internal services do use Http polling quite frequently so we are trying to multiplex different service APIs through a single WebSocket connection. Every service will essentially have their own "namespace" within centrifuge. 

When a user disconnects ungracefully due to a loss of connection the services have asked if they could be notified. What that means for us is we need to see the channels the user is subscribed to on disconnect for specific disconnect reasons. Our idea was iterate through the channels the user was subscribed to and send a single notification per "namespace" that the user is subscribed to. 

For example:

User1 is subscribed to: [namespace1:channel1, namespace2:channel1, namespace1: channel2]

When he disconnects due to a specific reason we want to notify  namespace1 and namespace2 so they can potentially react to the User losing a connection

**Comment by FZambia** [comment]
@patrick-janeiro @EmandM thanks for providing details!

So do I understand right that adding sth like this will suffice your use case? 

```diff
--- a/events.go
+++ b/events.go
@@ -105,6 +105,11 @@ type DisconnectEvent struct {
        // is nil then this means that client disconnected normally and connection
        // closing was initiated by client side.
        Disconnect *Disconnect
+       // ChannelsBeforeDisconnect is a snapshot map with all channels client was subscribed
+       // to before disconnect happened. At the time of issuing DisconnectEvent Client
+       // already unsubscribed from these channels and UnsubscribeEvent for each channel
+       // in ChannelsBeforeDisconnect already fired.
+       ChannelsBeforeDisconnect map[string]ChannelContext
 }
```

Keys are channels, `ChannelContext` is a struct with no exposed fields (i.e. we exposing [this](https://github.com/centrifugal/centrifuge/blob/d57e7e003accb56c608c8088af8a5df0fba14b06/client.go#L785) map actually).

Do you also need to distinguish between user-initiated unsubscribe events and unsubscribe events caused by disconnection logic? (I mean `UnsubscribeEvent` in `OnUnsubscribe`).

Also one thing that may be not obvious and should be better documented in Centrifuge is that Centrifuge can't differentiate disconnections caused by client side called `centrifugeClient.Disconnect()` method of connector library or connection closed due to client network problem – in both cases *Disconnect in DisconnectEvent will be `nil` - i.e. all client disconnects are the same for a server at the moment (and I'd like to keep it this way since it can be hard to distinguish between two in different transports). But since you are mentioned `unexpected` connection close I decided it's important to emphasise. Is this fine for you?

**Comment by patrick-janeiro** [comment]
@FZambia  That would be more then sufficient :) 

We don't currently need to know the why the unsubscribe was caused yet since we only care about firing the event to the namespace owner on actual disconnect.

I'll bring that up to the team, you are correct to emphasize this as I believe that currently our logic assumes that if the `*Disconnect` pointer is `nil` then something went wrong. But I believe we have ways to get around this. 

Would you like us to put a PR for the `ChannelsBeforeDisconnect`?

**Comment by FZambia** [comment]
Just tried to implement this and realized that it's not that straightforward as it seemed at first glance.

1. This map will escape to the heap in this case.
2. This map may contain channels in `subscribing` state (subscriptions on the fly) – thus we need to create a copy of it which is not a good thing.
3. If you decide to handle unsubscribe events at some point you will need to distinguish between two anyway (caused by a client or caused by disconnect in a progress).

A possible alternative which seems cheap is:

```diff
type UnsubscribeEvent struct {
        Channel string
        // ServerSide set to true for server-side subscription unsubscribe events.
        ServerSide bool
+       // Disconnecting set to true when UnsubscribeEvent issued due to disconnect
+       // process in a progress.
+       Disconnecting bool
+       // Disconnect can be set when Disconnecting is true. If connection close was
+       // initiated by a server it will contain Disconnect object, otherwise will if
+       // client connection lost or closed normally from a client side Disconnect is nil.
+       Disconnect *Disconnect
 }
```

Maybe instead of `Disconnecting` flag it's better to introduce `UnsubscribeSource` enum: `UnsubscribeSourceServer`, `UnsubscribeSourceClient`, `UnsubscribeSourceDisconnect`. `Disconnect` already escapes to the heap so no additional allocations should happen here.

Sorry – the code base is pretty huge so it's sometimes hard to notice all the caveats. What do you think about this approach?

**Comment by patrick-janeiro** [comment]
I'm actually really glad you brought this approach up, I prefer the enumeration because it provides a bit more granularity when it comes to why the user is unsubscribed. I do have a couple of questions though. 

With the code above we will execute the callback on every channel. This indicates that in my use-case above we would actually send two "notifications" to our `namespace1`.  While technically the above approach works and we will be able to send a notification it also implies means we need to figure out how to dedup the notifications per namespace. 

This however does cover our use-case in terms of functionality.

**Comment by FZambia** [comment]
I think it will be still possible to batch notifications:

```go
node.OnConnect(func(client *centrifuge.Client) {
	var channelsBeforeDisconnect []string

	client.OnUnsubscribe(func(e centrifuge.UnsubscribeEvent) {
		// Add to batch.
		if e.Disconnecting { // Or checking enum if we go with it.
			channelsBeforeDisconnect = append(channelsBeforeDisconnect, e.Channel)
		}
	})

	client.OnDisconnect(func(e centrifuge.DisconnectEvent) {
		// Send notifications here.
		log.Printf("channels before disconnect: %#v", channelsBeforeDisconnect)
	})
})
```

This way Centrifuge users who don't need this functionality won't pay extra allocs if disconnect handler set. What do you think?

**Comment by EmandM** [comment]
This works great for us. We think the Enum would be best, but otherwise this is a great solution. Thank you!

**Comment by FZambia** [comment]
Opened pull request #213 – could you plz check it out?

**Comment by FZambia** [comment]
Available in [v0.18.9](https://github.com/centrifugal/centrifuge/releases/tag/v0.18.9) 🎉
